refer/sortbib.1
.\" Copyright (c) 1983 Regents of the University of California.
.\" All rights reserved.
.\"
.\" This code contains changes by
.\"      Gunnar Ritter, Freiburg i. Br., Germany, 2005. All rights reserved.
.\"
.\" Conditions 1, 2, and 4 and the no-warranty notice below apply
.\" to these changes.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowedgement:
.\" 	This product includes software developed by the University of
.\" 	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"   Redistributions of source code and documentation must retain the
.\"    above copyright notice, this list of conditions and the following
.\"    disclaimer.
.\"   Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"   All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\"   Neither the name of Caldera International, Inc. nor the names of
.\"    other contributors may be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE
.\" LIABLE FOR ANY DIRECT, INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" Sccsid @(#)sortbib.1b	1.4 (gritter) 12/12/05
.\"
.\"	from 4.3BSD-Tahoe sortbib.1	6.1 (Berkeley) 4/29/85
.\"
.\".TH SORTBIB 1 "April 29, 1985"
.TH SORTBIB 1 "12/12/05" "Heirloom Documentation Tools" "BSD System Compatibility"
.UC 5
.SH NAME
sortbib \- sort bibliographic database
.SH SYNOPSIS
\fBsortbib\fR [\fB\-s\ \fIKEYS\fR] \fIdatabase\fR ...
.SH DESCRIPTION
.I Sortbib
sorts files of records containing
.I refer
key-letters by user-specified keys.
Records may be separated by blank lines,
or by \&.[ and \&.] delimiters,
but the two styles may not be mixed together.
This program reads through each
.I database
and pulls out key fields, which are sorted separately.
The sorted key fields contain the file pointer,
byte offset, and length of corresponding records.
These records are delivered using disk seeks and reads, so
.I sortbib
may not be used in a pipeline to read standard input.
.PP
By default,
.I sortbib
alphabetizes by the first %A and the %D fields,
which contain the senior author and date.
The
.B \-s
option is used to specify new
.IR \s-1KEYS\s0 .
For instance,
.BR \-s ATD
will sort by author, title, and date,
while
.BR \-s A+D
will sort by all authors, and date.
Sort keys past the fourth are not meaningful.
No more than 16 databases may be sorted together at one time.
Records longer than 4096 characters will be truncated.
.PP
.I Sortbib
sorts on the last word on the %A line,
which is assumed to be the author's last name.
A word in the final position, such as ``jr.'' or ``ed.'',
will be ignored if the name beforehand ends with a comma.
Authors with two-word last names or unusual constructions
can be sorted correctly by using the
.I nroff
convention ``\e0'' in place of a blank.
A %Q field is considered to be the same as %A,
except sorting begins with the first, not the last, word.
.I Sortbib
sorts on the last word of the %D line, usually the year.
It also ignores leading articles (like ``A'' or ``The'')
when sorting by titles in the %T or %J fields;
it will ignore articles of any modern European language.
If a sort-significant field is absent from a record,
.I sortbib
places that record before other records containing that field.
.SH SEE ALSO
refer(1), addbib(1), roffbib(1), indxbib(1), lookbib(1)
.SH NOTES
Records with missing author fields
should probably be sorted by title.
.PP
Written by
Greg Shenaut & Bill Tuthill.


CHANGES
2014-10-29
    Documentation revised by Pierre-Jean Fichet:  doc/fonts/
    changed to use free available OTF fonts.

2014-10-27
    col(1) bugfix: Avoid to access chars before string.

    troff: fdprintf changed back to libc fprintf.

    troff: 32 bit bug in register and macro allocation fixed (Bug
    reported by Anthony J. Bentley).

    nroff/troff made -Wall clean.

2014-10-22
    OpenSolaris col(1) added.

2014-10-06
    Drawing of filled polygone (\D'P ...'), filled circle (\D'C
    ...') and filled ellipse (\D'E ...') added.

2014-09-21
    *roff: Option -V for version information added.

2014-09-14
    Any version of roff (e.g. UNIX V7) does count \} as a macro
    argument regarding \n(.$.  This makes \n(.$ rather random if
    you cannot force users to put \} on a separate line *or* to
    put it behind the macro call and substract 1 from \n(.$.  To
    make \n(.$ meaningful again this bug has been fixed now, i.e.
    \} is not considered as a macro argument anymore.

2014-09-07
    Documentation revised by Pierre-Jean Fichet:  doc/just/ and
    doc/quickstart/ changed to use free available OTF fonts.

2014-09-03
    Bug reported by Pierre-Jean Fichet fixed:  If character '"'
    is used in a .char request it did not work in a .if request
    as the string compare delimiter.

2014-08-28
    Bug reported by Blake McBride fixed:  Very old versions of
    the mm macros inserted cut marks in the document for
    typesetters which had used endless paper rolls.  Although
    they are only of historical significance they had never been
    removed in the Solaris version of mm.  They are now switched
    of (via .rm )k).

2014-08-23
    Bugfix by Steffen Nurpmeso:  In .rm code a function return
    value is unchecked used as a pointer.

2014-08-20
    eqn(1) did not compile because of bison to BSD yacc differ-
    ences.  Sorry, had not tested on Linux this time...

2014-08-18
    Experimental bugfix: When .bp is directly followed by .. and
    .bp triggers a macro which calls a macro then document text
    is printed at that trap position.  This bug is also in other
    ditroff derivatives (e.g. DWB and Plan9).  It is difficult to
    track the bug to it's root cause so it is an experimental fix
    at the moment.

2014-08-17
    \fP after \fS did set the font before the previous font and
    not the previous one.

2014-08-16
    Bug reported by Pierre-Jean Fichet fixed:  If character "!"
    is used in a .char request it did not work in a .if request
    as the NOT operator.

2014-08-14
    The eqn parser used type int for all token values--even
    pointers.  This may work fine on 32 bit hardware but causes
    problems on 64 bit.  A new data type had been introduced to
    separate int tokens from pointers.

    Also a bug in the parser regarding movements (e.g. FWD) had
    been fixed.

2014-07-27
    A variable had been initilized after use.  This variable
    minus a literal value is used as an array index.  When data
    from the array is read an invalid memory segment is accessed.
    This had caused crashes on at least OpenBSD.


refer/sortbib.c
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved. The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

/*
 * Copyright (c) 1983-1988 by Sun Microsystems, Inc.
 * All Rights Reserved.
 */

/*	from OpenSolaris "sortbib.c	1.9	05/06/02 SMI"	*/

/*
 * Portions Copyright (c) 2005 Gunnar Ritter, Freiburg i. Br., Germany
 *
 * Sccsid @(#)sortbib.c	1.3 (gritter) 10/22/05
 */

#include <locale.h>
#include <stdio.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define	BUF BUFSIZ
#define	MXFILES 16

char tempfile[32];		/* temporary file for sorting keys */
int tmpfd = -1;
char *keystr = "AD";		/* default sorting on author and date */
int multauth = 0;		/* by default sort on senior author only */
int oneauth;			/* has there been author in the record? */

static void sortbib(FILE *, FILE *, int);
static void deliver(FILE **, FILE *);
static void parse(char *, char [][BUF]);
static int article(const char *);
static void eval(char *);
static void error(const char *);
static void onintr(int);
static int endcomma(char *);

int
main(int argc, char **argv)	/* sortbib: sort bibliographic database in place */
{
	FILE *fp[MXFILES], *tfp;
	int i;

	if (argc == 1)		/* can't use stdin for seeking anyway */
	{
		puts("Usage:  sortbib [-sKEYS] database [...]\n\
\t-s: sort by fields in KEYS (default is AD)");
		exit(1);
	}
	if (argc > 2 && argv[1][0] == '-' && argv[1][1] == 's')
	{
		/* if a key is specified use it, otherwise use default key */
		if (argv[1][2] != '\0')
			keystr = argv[1] + 2;
		eval(keystr);		/* evaluate A+ for multiple authors */
		argv++; argc--;
	}
	if (argc > MXFILES+1)	/* too many open file streams */
	{
		fprintf(stderr,
		"sortbib: More than %d databases specified\n",
		    MXFILES);
		exit(1);
	}
	for (i = 1; i < argc; i++)		/* open files in arg list */
		if ((fp[i-1] = fopen(argv[i], "r")) == NULL)
			error(argv[i]);
	strcpy(tempfile, "/tmp/SbibXXXXXX");	/* tempfile for sorting keys */
	if ((tmpfd = mkstemp(tempfile)) == -1)
		error(tempfile);

	close(tmpfd);
	if (signal(SIGINT, SIG_IGN) != SIG_IGN)	/* remove if interrupted */
		signal(SIGINT, onintr);
	if ((tfp = fopen(tempfile, "w")) == NULL) {
		unlink(tempfile);
		error(tempfile);
	}
	for (i = 0; i < argc-1; i++)		/* read keys from bib files */
		sortbib(fp[i], tfp, i);
	fclose(tfp);
	deliver(fp, tfp);	/* do disk seeks and read from biblio files */
	unlink(tempfile);
	return 0;
}

static int rsmode = 0;		/* record separator: 1 = null line, 2 = bracket */

static void
sortbib(FILE *fp, FILE *tfp, int i)	/* read records, prepare list for sorting */
{
	long offset, lastoffset = 0;		/* byte offsets in file */
	int length, newrec = 0, recno = 0;	/* reclen, new rec'd?, number */
	char line[BUF], fld[4][BUF];		/* one line, the sort fields */

	/* measure byte offset, then get new line */
	while (offset = ftell(fp), fgets(line, BUF, fp))
	{
		if (recno == 0)		/* accept record w/o initial newline */
			newrec = 1;
		if (line[0] == '\n')	/* accept null line record separator */
		{
			if (!rsmode)
				rsmode = 1;	/* null line mode */
			if (rsmode == 1)
				newrec = 1;
		}
		if (line[0] == '.' && line[1] == '[')	/* also accept .[ .] */
		{
			if (!rsmode)
				rsmode = 2;	/* bracket pair mode */
			if (rsmode == 2)
				newrec = 1;
		}
		if (newrec)		/* by whatever means above */
		{
			newrec = 0;
			length = offset - lastoffset;	/* measure rec len */
			if (length > BUF*8) {
				fprintf(stderr,
				"sortbib: record %d longer than %d (%d)\n",
					recno, BUF*8, length);
				unlink(tempfile);
				exit(1);
			}
			if (recno++)			/* info for sorting */
			{
				fprintf(tfp, "%d %ld %d : %s %s %s %s\n",
					i, lastoffset, length,
					fld[0], fld[1], fld[2], fld[3]);
				if (ferror(tfp)) {
					unlink(tempfile);
					error(tempfile);
				}
			}
			*fld[0] = *fld[1] = *fld[2] = *fld[3] = 0;
			oneauth = 0;		/* reset number of authors */
			lastoffset = offset;	/* save for next time */
		}
		if (line[0] == '%')	/* parse out fields to be sorted */
			parse(line, fld);
	}
	offset = ftell(fp);		/* measure byte offset at EOF */
	length = offset - lastoffset;	/* measure final record length */
	if (length > BUF*8)
	{
		fprintf(stderr,
		    "sortbib: record %d longer than %d (%d)\n",
		    recno, BUF*8, length);
		unlink(tempfile);
		exit(1);
	}
	if (line[0] != '\n')		/* ignore null line just before EOF */
	{
		fprintf(tfp, "%d %ld %d : %s %s %s %s\n",
			i, lastoffset, length,
			fld[0], fld[1], fld[2], fld[3]);
		if (ferror(tfp)) {
			unlink(tempfile);
			error(tempfile);	/* disk error in /tmp */
		}
	}
}

static void
deliver(FILE **fp, FILE *tfp)	/* deliver sorted entries out of database(s) */
{
	char str[BUF], buff[BUF*8];	/* for tempfile & databases */
	char cmd[80];			/* for using system sort command */
	long int offset;
	int i, length;

	/* when sorting, ignore case distinctions; tab char is ':' */
	sprintf(cmd, "sort +4f +0n +1n %s -o %s", tempfile, tempfile);
	if (system(cmd) == 127) {
		unlink(tempfile);
		error("sortbib");
	}
	tfp = fopen(tempfile, "r");
	while (fgets(str, sizeof (str), tfp))
	{
		/* get file pointer, record offset, and length */
		if (sscanf(str, "%d %ld %d :", &i, &offset, &length) != 3)
			error("sortbib: sorting error");
		/* seek to proper disk location in proper file */
		if (fseek(fp[i], offset, SEEK_SET) == -1) {
			unlink(tempfile);
			error("sortbib");
		}
		/* read exactly one record from bibliography */
		if (fread(buff, sizeof (*buff), length, fp[i]) == 0) {
			unlink(tempfile);
			error("sortbib");
		}
		/* add newline between unseparated records */
		if (buff[0] != '\n' && rsmode == 1)
			putchar('\n');
		/* write record buffer to standard output */
		if (fwrite(buff, sizeof (*buff), length, stdout) == 0) {
			unlink(tempfile);
			error("sortbib");
		}
	}
}

static void
parse(char *line, char fld[][BUF])	/* get fields out of line, prepare for sorting */
{
	char wd[8][BUF/4];
	int n, i, j;

	for (i = 0; i < 8; i++)		/* zap out old strings */
		*wd[i] = 0;
	n = sscanf(line, "%s %s %s %s %s %s %s %s",
		wd[0], wd[1], wd[2], wd[3], wd[4], wd[5], wd[6], wd[7]);
	for (i = 0; i < 4; i++)
	{
		if (wd[0][1] == keystr[i])
		{
			if (wd[0][1] == 'A')
			{
				if (oneauth && !multauth)	/* no repeat */
					break;
				else if (oneauth)		/* mult auths */
					strcat(fld[i], "~~");
				if (!endcomma(wd[n-2]))		/* surname */
					strcat(fld[i], wd[n-1]);
				else {				/* jr. or ed. */
					strcat(fld[i], wd[n-2]);
					n--;
				}
				strcat(fld[i], " ");
				for (j = 1; j < n-1; j++)
					strcat(fld[i], wd[j]);
				oneauth = 1;
			} else if (wd[0][1] == 'D') {
				strcat(fld[i], wd[n-1]);	/* year */
				if (n > 2)
					strcat(fld[i], wd[1]);	/* month */
			} else if (wd[0][1] == 'T' || wd[0][1] == 'J') {
				j = 1;
				if (article(wd[1]))	/* skip article */
					j++;
				for (; j < n; j++)
					strcat(fld[i], wd[j]);
			} else  /* any other field */
				for (j = 1; j < n; j++)
					strcat(fld[i], wd[j]);
		}
		/* %Q quorporate or queer author - unreversed %A */
		else if (wd[0][1] == 'Q' && keystr[i] == 'A')
			for (j = 1; j < n; j++)
				strcat(fld[i], wd[j]);
	}
}

static int
article(const char *str)		/* see if string contains an article */
{
	if (strcmp("The", str) == 0)	/* English */
		return (1);
	if (strcmp("A", str) == 0)
		return (1);
	if (strcmp("An", str) == 0)
		return (1);
	if (strcmp("Le", str) == 0)	/* French */
		return (1);
	if (strcmp("La", str) == 0)
		return (1);
	if (strcmp("Der", str) == 0)	/* German */
		return (1);
	if (strcmp("Die", str) == 0)
		return (1);
	if (strcmp("Das", str) == 0)
		return (1);
	if (strcmp("El", str) == 0)	/* Spanish */
		return (1);
	if (strcmp("Den", str) == 0)	/* Scandinavian */
		return (1);
	return (0);
}

static void
eval(char *keystr)		/* evaluate key string for A+ marking */
{
	int i, j;

	for (i = 0, j = 0; keystr[i]; i++, j++)
	{
		if (keystr[i] == '+')
		{
			multauth = 1;
			i++;
		}
		if (keystr[i] == 0)
			break;
		keystr[j] = keystr[i];
	}
	keystr[j] = 0;
}

static void
error(const char *s)		/* exit in case of various system errors */
{
	perror(s);
	exit(1);
}

static void
onintr(int unused)		/* remove tempfile in case of interrupt */
{
	fprintf(stderr, "\nInterrupt\n");
	unlink(tempfile);
	exit(1);
}

static int
endcomma(char *str)
{
	int n;

	n = strlen(str) - 1;
	if (str[n] == ',')
	{
		str[n] = '\0';
		return (1);
	}
	return (0);
}
